joomla/joomla-cms
# [com_fields] Feature components in component manifest

### Summary of Changes
We have some components in Joomla such as `com_categories` which are used across other extension. Another such extension recently added is `com_fields`. I call these as feature extensions, as they add features to other extensions with seamless integration.

Now that said, many other core as well as 3PD extensions makes use of them. Currently we do not have a way to tell whether a given component say `com_sample` uses these feature components or not. Similar issue was observed when creating a list of components which uses `com_categories` and we have no straight way to look at. Thankfully, @infograf768 came up with a best possible _workaround_ there. 

Here I am suggesting a straight path that can be used henceforth to identify all existing or possibly upcoming feature extension usage in other extensions. This is not just useful to create a menu link, rather when this information is made available on can think of various other uses.

### Prototype
Let us assume we have a component `com_sample` that uses `com_categories` and `com_fields` both. Currently as Joomla core or any other extension than `com_sample` itself, we do not have a clear way to know that.

I propose to update the install manifests structure for components as below.

**Current format**

```xml
<extension type="component" version="3.0" method="upgrade">
	<name>COM_SAMPLE</name>
	...
	<files folder="site">
		...
	</files>
	<administration>
		...
	</administration>
	...
</extension>
```

**Proposed Addition**

```xml
<extension type="component" version="3.0" method="upgrade">
	<name>COM_SAMPLE</name>
	...
	<files folder="site">
		...
	</files>
	<administration>
		...
	</administration>
	<features>
		<feature type="com_categories" name="com_sample" label="COM_SAMPLE_CATEGORIES_MAIN"/>
		<feature type="com_categories" name="com_sample.foo" label="COM_SAMPLE_CATEGORIES_FOO"/>
		<feature type="com_fields" name="com_sample" label="COM_SAMPLE_FIELDS_MAIN"/>
		<feature type="com_fields" name="com_sample.bar" label="COM_SAMPLE_FIELDS_BAR"/>
	</features>
	...
</extension>
```

This way we can now tell just by reading the component manifest that:
1. `com_sample` uses `com_categories` for two category types: `main` and `foo`
2. `com_sample` uses `com_fields` for two field types: `main` and `bar`

The format is:

```xml
<features>
	<feature type="FEATURE_COMPONENT_NAME" label="DISPLAY_NAME">
		
	</feature>
</features>
```

I have updated the core components manifests with this structure.

### Test Instruction
1. Create a custom admin menu and add a new menu item
2. Choose menu item type `Categories > List All Categories` and make sure that in the _Choose a Component_ dropdown the `Articles`, `Banners`, `Contact`, `Newsfeeds` and `Users` are listed.
3. Choose menu item type `Fields > List All Fields` and make sure that in the _Choose a Component_ dropdown the `Articles`, `Contact`, and `Users` are listed.
4. Install the attached `com_sample` and see if the above lists are updated to include a new entry `Sample` as well.
5. Make sure all those links created this way, works.



### Documentation Changes Required
Add this in manifest documentation.

[com_sample.zip](https://github.com/joomla/joomla-cms/files/752261/com_sample.zip)

Pinging @infograf768 @rdeutz
